### PR TITLE
Add username_type option

### LINKED ISSUE #3956: Add option for replacing username with email in User model
Ref #3484 

I'm not opposed to having an option to create a user model with only the email field and without a username field. I don't think a package (i.e. django-improved-user) requires that. I see myself always needing to alter the admin, user model, user manager, allauth configuration, user forms, the user migration file, drf api, `reverse` code, tests, and views. On the other hand, I also have seen at times that I do need a username field in addition to email (i.e. the code we see today).

Adding django-improved-user would still require changing the majority of the aforementioned code in cookiecutter-django. Propose adding an option in the template for email as username?

### PR DESCRIPTION
**Pull request by Andrew-Chen-Wang**
## Description

Fixes #3956

The type of username you want to use in the project. This can be either``username`` or ``email``. If you choose ``username``, the ``email`` field will be included. If you choose ``email``, the ``username`` field will be excluded. It is best practice to always include an email field, so there is no option for having just the ``username`` field.

Checklist:

- [X] I've made sure that tests are updated accordingly (especially if adding or updating a template option)
- [X] I've updated the documentation or confirm that my change doesn't require any updates

## Rationale

I encounter needing to exclude username field several times.

### CODE REVIEW

**Comment by Andrew-Chen-Wang** [comment]
stuck trying to fix all the linting problems. Will come back later.

File: {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/users/migrations/0001_initial.py [hunk 1]
```diff
@@ -40,6 +42,7 @@ class Migration(migrations.Migration):
                         verbose_name="superuser status",
                     ),
                 ),
+                {%- if cookiecutter.username_type == "email" -%}
```
[the remark this answers is not in the document]
Comment (Andrew-Chen-Wang) [reply {{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/users/migrations/0001_initial.py:45]:
```suggestion
                {%- if cookiecutter.username_type == "username" -%}
```

**Comment by Andrew-Chen-Wang** [comment]
Thanks @sorasful for the review!

**Comment by DonExo** [comment]
Let's get this reviewed and merged this please. It is very much a standard nowadays to go ahead with email instead of username(s).

**Review by DonExo — APPROVED** [review]

**Comment by browniebroke** [comment]
> Let's get this reviewed and merged this please

@DonExo I assume it wasn't intentional, but FYI that came across as an order from a manager to its subordinates. You're welcome to become "my boss" and give me a financial push by hitting the sponsor button at the top of the page 😄  Otherwise, this project is run by volunteers, on their free time.

In terms of this pull request, you're right that it's a pretty common use case, and it would indeed be nice to get this merged! However, as it stands, this change is blocked, and to unblock it, here is what needs to happen:

- The build on CI should be green
- Resolve merge conflicts
- Write tests for the `UserManager` when `email` is selected. 
- Ideally, cover this option in the e2e tests (add a job under the `bare` section of the CI jobs)
- Do a bit of manual testing to make sure it works on prod

Honestly, the hardest part is already done, this is a pretty solid PR already as it stands, thanks @Andrew-Chen-Wang

**Comment by Andrew-Chen-Wang** [comment]
Let me know if you're unable to edit the branch @browniebroke No time to fix it up, sorry but thanks so much!

**Comment by browniebroke** [comment]
Thanks for putting this together @Andrew-Chen-Wang. I had to do it after generation recently on a new project and this PR even unmerged saved me a ton of time. Hopefully it'll save even more time by being an option now 🚀

**Pull request merged**